Fix write-exec-ETXTBSY race

DrJosh9000 opened the pull request:
I'm reluctant to fix bintest, but the number of tests we have in the agent using bintest is still high, and the flakes are annoying.

There is a bug in Go, or a design flaw in Linux and some other/older Unices (which one depends on your point of view). When writing and then executing binaries in a multithreaded program, there is a race condition that causes the exec to fail with `ETXTBSY`. See https://github.com/golang/go/issues/22315.

This shows up as a flake in some of the agent tests, as exit code -1 on running the bintest client binary (the underlying error is `text file busy`, which is normally uploaded in the job log, but is inaccessible to the tests).

A solution is to build the client binary directly into the cache, and replace `copyFile` with symlinking. This works because [symlink(2)](https://man7.org/linux/man-pages/man2/symlink.2.html) is a single syscall, and at no point requires the process to have a file descriptor open with write on the symlink (which would temporarily leak into forked processes). This has the added benefit of being nicer to the filesystem (symlinks are small compared with Go binaries).

Aside: why doesn't the initial binary compile have this problem? Because the exec of `go build` is a subprocess, its file descriptors don't sneak back into the parent process. (So another solution to the race is to do file copying in a subprocess).

The agent tests pass*, which I verified in a Linux container and a `go.work` configuration (`cd ..` into a common superdirectory, `go work init`, `go work use (path to agent repo)`, `go work use (path to bintest repo)`).

*this was a slight exaggeration when I wrote it, but not now. (Had to install Ruby, copy .buildkite/build/ssh.conf to ~/.ssh/config, and fix two more edge cases.)

DrJosh9000 committed:
Update go version in go.mod

DrJosh9000 committed:
Replace ioutil with io and os equivs

DrJosh9000 committed:
Simplify IsCached

DrJosh9000 committed:
Add test that detects write-exec race

DrJosh9000 committed:
Simplify temp subdir handling

DrJosh9000 committed:
Replace copyFile with symlinking

triarius commented on compiler.go:
```diff
@@ -86,48 +85,47 @@ func compileClient(dest string, vars []string) error {
 
-	if _, err := os.Lstat(dir); os.IsNotExist(err) {
-		_ = os.Mkdir(filepath.Join(wd, dir), 0700)
+	if err := os.MkdirAll(dir, 0700); err != nil {
```
```suggestion
	if err := os.MkdirAll(dir, 0o700); err != nil {
```

triarius commented on compiler.go:
```diff
@@ -86,48 +85,47 @@ func compileClient(dest string, vars []string) error {
-			_ = os.RemoveAll(dir)
-			return err
-		}
+	if err := os.WriteFile(f, []byte(clientSrc), 0500); err != nil {
```
```suggestion
	if err := os.WriteFile(f, []byte(clientSrc), 0o500); err != nil {
```

triarius commented on testutil/testutil.go:
```diff
@@ -39,13 +39,13 @@ func (t *TestingT) Copy(dst *testing.T) {
 
 	batchfile := filepath.Join(tmpDir, name)
-	err = ioutil.WriteFile(batchfile, []byte(strings.Join(lines, "\r\n")), 0600)
+	err = os.WriteFile(batchfile, []byte(strings.Join(lines, "\r\n")), 0600)
```
```suggestion
	err = os.WriteFile(batchfile, []byte(strings.Join(lines, "\r\n")), 0o600)
```

DrJosh9000 committed:
exec.LookPath for non-absolute args[0]

Some agent tests invoke binaries via PATH resolution. os.Executable worked before, because the invoked path was always an executable (that bintest copied into place). Now that the invoked path is always a symlink, os.Executable is wrong and exec.LookPath is right.

triarius reviewed (approved):
Amazing work! Another blog post? Though this is a bit of a niche issue...

My only comment is that we can do a drive by cleanup of the archaic octal notation while we are in the area. But this is very much non-blocking.

DrJosh9000 committed:
Use newer octal notation

Co-authored-by: Narthana Epa <narthana.epa@gmail.com>

DrJosh9000 commented:
> Amazing work!

😊

> Another blog post? Though this is a bit of a niche issue...

Mmmmmmmaybe...

DrJosh9000 committed:
replaceSymlink

DrJosh9000 committed:
Fix for invocation via relative path with /
